### PR TITLE
chore: @microsoft/generator-bot-adaptive happy path unit tests

### PR DESCRIPTION
**Pull request by peterinnesmsft**
### Purpose
Partially addresses #1084.

To help improve overall code quality and keep it at an acceptable bar, we need to add automated unit test coverage for our generators.

This PR is the first step in adding unit test coverage, starting with our base generator, @microsoft/generator-bot-adaptive. The goal of this initial PR is to not add a fully exhaustive set of tests, but rather enough to tease out the pattern so that other generators can follow suit, but identify if we are happy with the basic style, structure and approach.

Keeping in line with the JS SDK, I am making use of the Mocha test framework, as well as Yeoman's own yeoman-test package with provides capabilities for easily executing and verifying the outcome of a generator.

This PR adds happy path coverage for our four key permutations:
 - .NET Web App
 - .NET Functions
 - JS Web App
 - JS Functions

### Changes
- Update eslintrc to support mocha and ES2017 environments so that it can correctly identify expected syntax.
- Added new test script command to the package.json, as well as add devDependencies for ejs, mocha and yeoman-test.
- Add unit test files and helpers to tests/ in package root.

### Tests
This PR adds 53 unit tests for @microsoft/generator-bot-adaptive, grouped into the four scenarios as described above - each test asserts the presence and contents of each expected file in the outlined scenarios.

### Feature Plan
Two main elements need to be addressed further:
- More test coverage for handling edge cases, asserting no additional files have been added, specification of options as would be expected from our child generators, and more general error validation.
- Figure out how to incorporate these into our CI/CD pipelines. I would expect these are added to both our PR and Daily pipelines, but open to discuss and land elsewhere if people feel strongly otherwise.

#minor

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -1,6 +1,7 @@
 {
   "env": {
-    "es6": true,
+    "es2017": true,
```
Comment (joshgummersall) [inline .eslintrc.json:3]: Why do we need this?

File: generators/generator-bot-adaptive/package.json [hunk 2]
```diff
@@ -20,7 +20,8 @@
     "url": "https://github.com/Microsoft/botframework-components.git"
   },
   "scripts": {
-    "lint": "eslint -c .eslintrc.json ."
+    "lint": "eslint -c .eslintrc.json .",
+    "test": "mocha tests"
```
Comment (joshgummersall) [inline generators/generator-bot-adaptive/package.json:24]: Small note, but `mocha` convention defaults to `test`. If it's not too annoying, the `tests` directory should be named `test` and then this command can simply be `mocha`.

File: generators/generator-bot-adaptive/tests/suites/js-functions.test.js [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+  runResult.assertFileContent(filePath, content);
+});
+
+it(`should create file ${path.join(botName, 'package.json')}`, function () {
+  const filePath = path.join(botName, 'package.json');
+  const content = {
+    name: botName,
+    private: true,
+    scripts: {
+      dev: 'cross-env NODE_ENV=dev func start',
```
Comment (joshgummersall) [inline generators/generator-bot-adaptive/tests/suites/js-functions.test.js:102]: I wonder if we should be a bit less strict about these expectations - I suppose we can adjust this if it gets annoying having to update commands or dependency versions in two places.

File: generators/generator-bot-adaptive/tests/generator.test.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+function importTests(name, path) {
+  describe(name, function () {
+    require(path);
+  });
+}
+
+describe('generator-bot-adaptive', function () {
```
Comment (joshgummersall) [inline generators/generator-bot-adaptive/tests/generator.test.js:10]: I think `mocha` should recursively discover everything in the `test` directory. This pattern of registering tests in a single file is a little unusual. Generally, each test would `describe` its own hierarchy of tests.

All four tests are fairly common now, but we might want to add other language-specific tests. In my (humble) opinion, this pattern doesn't have enough shared utility to warrant the non-standard structure that may impact our ability to iterate on the tests in the future.

**Review by joshgummersall — COMMENTED** [review]
Nice! I would probably decouple the tests (remove the `importTests` helper and enumerate the full `describe` structure in each test) to have more flexibility in the future. Otherwise, this looks great.

File: generators/generator-bot-adaptive/package.json (hunk 2 above)
Comment (joshgummersall) [reply generators/generator-bot-adaptive/package.json:24]: You may need to add something like: 
```json
{
  ...,
  "mocha": {
    "recursive": true
  },
  ...
}
```
to properly discover all the tests as well. I'd also suggest adding `checkLeaks`, `exit`, `forbidOnly`, and `forbidPending`.

**Comment by tracyboehrer** [comment]
Would be helpful to perform a build for Dotnet.  This will catch cases where we've induced compile time errors in the template, which we've been prone to do here an there.

**Comment by tracyboehrer** [comment]
Where do we stand on this?  Josh G's suggestion for discovering the tests seem reasonable.  Is there a reason we wouldn't want to do that?

**Comment by tracyboehrer** [comment]
It would be helpful to make sure expected template tags aren't trampled as well.  For example, the namespace for the cs files are updated to the bot name, etc...

File: .eslintrc.json (hunk 1 above)
Comment (peterinnesmsft) [reply .eslintrc.json:3]: Discussed offline. This was specifically added to support async function () syntax in the before hook for each suite of tests, otherwise eslint would throw an error.

File: generators/generator-bot-adaptive/tests/generator.test.js (hunk 4 above)
Comment (peterinnesmsft) [reply generators/generator-bot-adaptive/tests/generator.test.js:10]: I've moved all tests up to a `test` dir and removed this file.

File: generators/generator-bot-adaptive/tests/suites/js-functions.test.js (hunk 3 above)
Comment (peterinnesmsft) [reply generators/generator-bot-adaptive/tests/suites/js-functions.test.js:102]: Spoke offline - going to leave as is for now. If this becomes problematic, it can be addressed in the future.

File: generators/generator-bot-adaptive/package.json [hunk 5]
```diff
@@ -21,14 +21,14 @@
...
-    "test": "mocha tests"
+    "test": "mocha"
   },
   "dependencies": {
     "globby": "^11.0.2",
     "normalize-path": "^3.0.0",
     "runtypes": "~5.1.0",
     "uuid": "^8.3.1",
-    "yeoman-generator": "^1.1.1"
+    "yeoman-generator": "^2.0.5"
```
Comment (joshgummersall) [inline generators/generator-bot-adaptive/package.json:31]: Is the the furthest version bump we can take before there are breaking API changes?

Comment (peterinnesmsft) [reply generators/generator-bot-adaptive/package.json:31]: @joshgummersall it's not, but it's consistent with the version that is being used in the child generators. Per our discussion, figured dependency version bump should come in a different PR. Does that work?

Comment (joshgummersall) [reply generators/generator-bot-adaptive/package.json:31]: Ah yep, consistency is good. I figured everything was on the older version already. This is a major version bump though, is there anything like a changelog I can triple check to make sure all the adaptive bot Yeoman code is still correct?

Comment (joshgummersall) [reply generators/generator-bot-adaptive/package.json:31]: Just looking at https://github.com/yeoman/generator/releases/tag/v2.0.0 now, seems fine to me. We never used the functional extends pattern, nor do we use the install methods.

**Review by joshgummersall — APPROVED** [review]
